### PR TITLE
Feat(optimizer)!: canonicalize table aliases

### PR DESCRIPTION
**Pull request by georgesittas**
This PR introduces a new `qualify` flag: `canonicalize_table_aliases`. This flag controls whether the `qualify_tables` rule should use canonical aliases (`_0`, `_1`, ...) for all sources instead of preserving table names.

### CODE REVIEW

File: sqlglot/optimizer/qualify_tables.py [hunk 1]
```diff
@@ -74,6 +77,32 @@ def _qualify(table: exp.Table) -> None:
             if isinstance(node, exp.Table) and node.name not in cte_names:
                 _qualify(node)
 
+    canonical_aliases: t.Dict[str, str] = {}
+
+    def _set_alias(
```
Comment (georgesittas) [inline sqlglot/optimizer/qualify_tables.py:76]: The purpose of this helper is to consolidate alias-setting logic into one place. We previously used to repeat `exp.alias_` calls in various places that you can find below in the diff. Moreover, we were also slightly inconsistent in the way we did that. Its arguments are needed to mimic the previous control flow.

A helpful side-effect of this is that it makes storing the old -> new alias mapping easy.

File: sqlglot/optimizer/qualify_tables.py [hunk 2]
```diff
@@ -18,6 +18,7 @@ def qualify_tables(
     catalog: t.Optional[str | exp.Identifier] = None,
     on_qualify: t.Optional[t.Callable[[exp.Expression], None]] = None,
     dialect: DialectType = None,
+    canonicalize: bool = False,
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_tables.py]: i think you'll want this to be different other it may clash with canonacilazie in qualify_columns if we add that later

File: sqlglot/optimizer/qualify_tables.py [hunk 3]
```diff
@@ -39,13 +40,15 @@ def qualify_tables(
...
+        canonicalize: Whether to use canonical aliases (_0, _1, ...) for all sources
+            instead of preserving table names. Defaults to False.
 
     Returns:
         The qualified expression.
     """
     dialect = Dialect.get_or_raise(dialect)
 
-    alias_sequence = name_sequence("_q_")
+    alias_sequence = name_sequence("_" if canonicalize else "_q_")
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_tables.py]: this doesn't need to be changed, you can keep it _q_ or make it _0

File: sqlglot/optimizer/qualify_tables.py [hunk 4]
```diff
@@ -74,6 +77,32 @@ def _qualify(table: exp.Table) -> None:
...
+    ) -> None:
+        alias = expression.args.get("alias") or exp.TableAlias()
+
+        if canonicalize:
+            new_alias_name = next_alias_name()
+            canonical_aliases[alias.name or target_alias or ""] = new_alias_name
+        elif not alias.name:
+            new_alias_name = target_alias or next_alias_name()
+            if normalize:
+                new_alias_name = normalize_identifiers(new_alias_name, dialect=dialect).name
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_tables.py:91]: why is there normalization going on here? normalization is already done in normalize step and next_alias_name is already normalized

File: tests/test_optimizer.py [hunk 5]
```diff
@@ -240,6 +240,77 @@ def test_isolate_table_selects(self):
         )
 
     def test_qualify_tables(self):
+        self.assertEqual(
```
Comment (tobymao) [inline tests/test_optimizer.py:247]: maybe let's put these into qualify_tables.sql with the option turned on

File: sqlglot/optimizer/qualify_tables.py (hunk 2 above)
Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py]: I can rename into `canonicalize_table_aliases` for consistency.

File: sqlglot/optimizer/qualify_tables.py (hunk 3 above)
Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py]: The problem with `_q` is that it can resolve to two different things: `_q` or `_Q`, depending on the dialect. Opted for the branch to avoid changing all the tests, but if you only want to keep that instead of `_q` then happy to change. Can also do it as a follow-up to avoid increasing the scope of this change.

Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py]: We won't need to normalize the alias sequence if we use `_number`, btw.

Comment (tobymao) [reply sqlglot/optimizer/qualify_tables.py]: let's change it all to just the number _1, breaking i know but might as well

File: sqlglot/optimizer/qualify_tables.py (hunk 4 above)
Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py:91]: I think it's because of this test case: https://github.com/tobymao/sqlglot/blob/df13a655646bd2ef5d8b4613670bb5fe48845b73/tests/fixtures/optimizer/optimizer.sql#L827-L860

If I remove normalization, the alias matches the table name, which is case-sensitive, and hence ends up being case-sensitive itself:

```diff
- FROM `bigquery-public-data.GooGle_tReNDs.TOp_TeRmS` AS `top_terms`
?                                                         ^^  ^ ^ ^
+ FROM `bigquery-public-data.GooGle_tReNDs.TOp_TeRmS` AS `TOp_TeRmS`
?                                                         ^^  ^ ^ ^
```

File: sqlglot/optimizer/qualify_tables.py (hunk 3 above)
Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py]: Changed

File: sqlglot/optimizer/qualify_tables.py (hunk 2 above)
Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py]: Renamed

File: sqlglot/optimizer/qualify_tables.py [hunk 6]
```diff
@@ -163,5 +174,13 @@ def _qualify(table: exp.Table) -> None:
...
 
                     column.set("table", table_alias.copy())
+            elif (
+                canonical_aliases
+                and table
+                and (canonical_table := canonical_aliases.get(table, "")) != column.table
+            ):
+                # Amend existing aliases, e.g. t.c -> _0.c if t is aliased to _0
+                column.set("table", exp.to_identifier(canonical_table))
+                pass
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_tables.py]: you don't need this pass right?

Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py]: yep thanks, left over from testing

Comment (tobymao) [reply sqlglot/optimizer/qualify_tables.py]: why do you only do this if the column doesn't have a db? if it does, doesn't mean you won't canonicalize it?

is there an edge case when a column could ambiguously reference a struct instead of a table and we don't know it yet because we haven't qualified columns?

File: tests/test_optimizer.py (hunk 5 above)
Comment (georgesittas) [reply tests/test_optimizer.py:247]: Will do shortly, wanna eyeball the optimizer test diff and then update these + look into adding more tests.

File: sqlglot/optimizer/qualify_tables.py (hunk 6 above)
Comment (georgesittas) [reply sqlglot/optimizer/qualify_tables.py]: Discussed this in Slack. The conclusion was that there exists a bug in main today too, where we incorrectly mutate multi-part column references, without checking whether they represent struct field lookups (i.e., `Dot` sequences).

We should look into this as a follow-up to this PR, to avoid increasing the current scope.

File: tests/test_optimizer.py (hunk 5 above)
Comment (georgesittas) [reply tests/test_optimizer.py:247]: cleaned this up

**Review by tobymao — APPROVED** [review]

**Comment by georgesittas** [comment]
This should be good to go. I'll see if I can increase the testing coverage a bit and then get it in.

**Comment by georgesittas** [comment]
Found a bug with correlated subqueries, i.e. when we have scopes w/ external columns. We need to respect lexical scoping in the canonical alias mapping probably. Going to fix and improve coverage.

File: sqlglot/optimizer/qualify_tables.py [hunk 7]
```diff
@@ -152,9 +169,11 @@ def _qualify(table: exp.Table) -> None:
 
         for table in scope.tables:
             if not table.alias and isinstance(table.parent, (exp.From, exp.Join)):
-                exp.alias_(table, table.name, copy=False, table=True)
+                _set_alias(table, canonical_aliases, target_alias=table.name)
+
+        for column in local_columns:
```
Comment (georgesittas) [inline sqlglot/optimizer/qualify_tables.py:174]: Note this change vs how we previously iterated over `scope.columns`. If we visited external columns in this loop, we would mutate their source too early, without having had the chance to alias outer scopes' sources.

Before making this change, the test case I added in `qualify_tables.sql` failed due to having columns qualified with `""`.

File: sqlglot/optimizer/qualify_tables.py [hunk 8]
```diff
@@ -74,7 +73,38 @@ def _qualify(table: exp.Table) -> None:
...
+
+        alias.set("this", exp.to_identifier(new_alias_name, quoted=quoted))
+        expression.set("alias", alias)
+
+        if scope:
+            scope.rename_source(None, new_alias_name)
+
     for scope in traverse_scope(expression):
+        local_columns = scope.local_columns
+        canonical_aliases: t.Dict[str, str] = {}
```
Comment (georgesittas) [inline sqlglot/optimizer/qualify_tables.py:106]: Auto-generated alias mappings are constructed per-scope to respect lexical scoping.

File: sqlglot/optimizer/scope.py [hunk 9]
```diff
@@ -383,12 +383,25 @@ def external_columns(self):
                 self._external_columns = [
                     c
                     for c in self.columns
-                    if c.table not in self.selected_sources
-                    and c.table not in self.semi_or_anti_join_tables
+                    if c.table not in self.sources and c.table not in self.semi_or_anti_join_tables
```
Comment (georgesittas) [inline sqlglot/optimizer/scope.py:386]: I ran this change against main and no tests failed. I don't see why we'd care _specifically_ about `selected_sources` here– @barakalon any ideas, or does this look reasonable to you as well?

The reason I needed to modify this was that, in order to access `external_columns` in the `qualify_tables` rule, we'd invoke `selected_sources` and this would fail in some cases due to "duplicate aliases", which in reality were simply `""`– since we haven't had the chance at that point to alias the derived tables.

Comment (barakalon) [reply sqlglot/optimizer/scope.py:386]: This is all about correlated subqueries - i.e. subqueries that reference columns in the outer scope.

`selected_sources` should be the table aliases from FROM/JOIN in the current scope.
So I think the intent here is "columns that reference a table that isn't selected in the current scope.... so it might come from the _outer_ scope"

`sources` might _include_ the table aliases from the outer scope? So I'm surprised this wouldn't break correlated subquery tests?

But this is one of the pieces of scope code that is coupled to the qualifier, so maybe you handled it in a different way?

Comment (georgesittas) [reply sqlglot/optimizer/scope.py:386]: > But this is one of the pieces of scope code that is coupled to the qualifier, so maybe you handled it in a different way?

Not aware of another change that is related to this. It just worked™ without much effort. 🤷‍♂️ 

> sources might include the table aliases from the outer scope? So I'm surprised this wouldn't break correlated subquery tests?

Does this approach concern you then? Do you have a hunch for what it could break?

Comment (barakalon) [reply sqlglot/optimizer/scope.py:386]: I mean, if tests pass.... 🚢
